_posts/2024-08-06-datafusion-python-udf-comparisons.md

```diff
@@ -0,0 +1,595 @@
+
+- Writing a pure python function to do your computation
+- Using the pyarrow libraries in python to accelerate your processing
+- Writing a UDF in rust and exposing it to python
+
+Additionally I will demonstrate two variants of this. The first will be nearly identical to the
+pyarrow library approach to simplicity of understanding how to connect the rust code to python. The
+second version we will do the iteration through the input arrays ourselves to give even greater
+flexibility to the user.
+
+Here are the two example use cases, taken from my own work but generalized.
+
+### Use Case 1: Scalar Function
+
+I have a DataFrame and a list of tuples that I’m interested in. I want to filter out the dataframe
+to only have values that match those tuples from certain columns in the dataframe. For example,
+suppose I have a table of sales line items. There are many columns, but I am interested in three: a
+part key, supplier key, and return status. I want only to return a dataframe with a specific
+combination of these three values.
+
+Probably the most ergonomic way to do this without UDF is to turn that list of tuples into a
+dataframe itself, perform a join, and select the columns from the original dataframe. If we were
+working in pyspark we would probably broadcast join the dataframe created from the tuple list since
+it is tiny. In practice, I have found that with some dataframe libraries performing a filter rather
+than a join can be significantly faster. This is worth profiling for your specific use case.
+
+### Use Case 2: Aggregate or Window Function
+
+I have a dataframe with many values that I want to aggregate. I have already analyzed it and
+determined there is a noise level below which I do not want to include in my analysis. I want to
+compute a sum of only values that are above my noise threshold.
+
+This can be done fairly easy without leaning on a User Defined Aggegate Function (UDAF). You can
+simply filter the dataframe and then aggregate using the built in `sum` function. Here, we
+demonstrate doing this as a UDF primarily as an example of how to write UDAFs. We will use the
+pyarrow compute approach.
+
+## Pure Python approach
+
+The fastest way (developer time, not code time) for me to implement the scalar problem solution
+was to do something along the lines of “for each row, check the values of interest contains that
+tuple”. I’ve published this as [an example](https://github.com/apache/datafusion-python/blob/main/examples/python-udf-comparisons.py)
+in the [datafusion-python repository](https://github.com/apache/datafusion-python). Here is an
+example of how this can be done:
+
+```python
+values_of_interest = [
+    (1530, 4031, "N"),
+    (6530, 1531, "N"),
+    (5618, 619, "N"),
+    (8118, 8119, "N"),
+]
+
+def is_of_interest_impl(
+    partkey_arr: pa.Array,
+    suppkey_arr: pa.Array,
+    returnflag_arr: pa.Array,
+) -> pa.Array:
```

Comment:
I think it might be helpful here to describe your problem a little bit. Say what `partkey_arr` is representing, and how it relates to your `values_of_interest` above.

Reply:
I had a smaller statement in the earlier section, but I've expanded it because it was easy to pass over.